### DISCUSSION TITLE
RADARSAT-2 processing with ISCE 2.6.2

### DISCUSSION THREAD

**Discussion by fdelgadodelapuente** [isce-framework/isce2#654] [General]
Dear ISCE developers,

ISCE fails when parsing RADASAT-2 files,

> 

fdelgado@prs:~/radarsat2/tiff$ unpackFrame_RSAT2.py -i RS2_OK125740_PK1100970_DK1047120_U16W2_20210203_094721_HH_SLC -o 20210203
This is the Open Source version of ISCE.
Some of the workflows depend on a separate licensed package.
To obtain the licensed package, please make a request for ISCE
through the website: https://download.jpl.nasa.gov/ops/request/index.cfm.
Alternatively, if you are a member, or can become a member of WinSAR
you may be able to obtain access to a version of the licensed sofware at
https://winsar.unavco.org/software/isce
effective PRF 3240.221191, original PRF 1620.110596
Traceback (most recent call last):
  File "/home/fdelgado/isce/isce2-2.6.2/contrib/stack/stripmapStack/unpackFrame_RSAT2.py", line 96, in <module>
    unpack(inps.RSATdir, inps.slcdir)
  File "/home/fdelgado/isce/isce2-2.6.2/install/isce/components/isceobj/Util/decorators.py", line 290, in use_api_decorator
    ret = func(*args,**kwargs)
  File "/home/fdelgado/isce/isce2-2.6.2/contrib/stack/stripmapStack/unpackFrame_RSAT2.py", line 52, in unpack
    obj.extractImage()
  File "/home/fdelgado/isce/isce2-2.6.2/install/isce/components/isceobj/Sensor/Radarsat2.py", line 319, in extractImage
    self.parse()
  File "/home/fdelgado/isce/isce2-2.6.2/install/isce/components/isceobj/Sensor/Radarsat2.py", line 113, in parse
    self.populateMetadata()
  File "/home/fdelgado/isce/isce2-2.6.2/install/isce/components/isceobj/Sensor/Radarsat2.py", line 196, in populateMetadata
    self.extractOrbit()
  File "/home/fdelgado/isce/isce2-2.6.2/install/isce/components/isceobj/Sensor/Radarsat2.py", line 262, in extractOrbit
    newOrb = orbExt.extendOrbit(tempOrbit)
  File "/home/fdelgado/isce/isce2-2.6.2/install/isce/components/isceobj/Orbit/OrbitExtender.py", line 218, in extendOrbit
    schOrb = self.getSCHOrbit(orbit, pegCoord, hdg)
  File "/home/fdelgado/isce/isce2-2.6.2/install/isce/components/isceobj/Orbit/OrbitExtender.py", line 117, in getSCHOrbit
    orbSch = stdproc.createOrbit2sch(averageHeight = peg.hgt)
  File "/home/fdelgado/isce/isce2-2.6.2/install/isce/components/stdproc/orbit/__init__.py", line 39, in createOrbit2sch
    from .Orbit2sch import Orbit2sch
  File "/home/fdelgado/isce/isce2-2.6.2/install/isce/components/stdproc/orbit/Orbit2sch.py", line 39, in <module>
    from stdproc.orbit import orbit2sch
ImportError: /home/fdelgado/isce/isce2-2.6.2/install/isce/components/stdproc/orbit/orbit2sch.abi3.so: undefined symbol: orbit2sch_

These files are called by the orbit extender. My understanding is that the state vectors provided by MDA are not long enough for processing the data with ISCE, so they need to be extended using either Hermite/Legendre polynomials. The practical solution was to copy some files from ISCE 2.2.0 through ISCE 2.6.2 and then recompile the software.

**mv isce2-2.6.2/components/stdproc/orbit/orbit2sch/src/SConscript isce2-2.6.2/components/stdproc/orbit/orbit2sch/src/SConscript_v2.6.2 

cp  isce-2.2.0/components/stdproc/orbit/orbit2sch/src/SConscript isce2-2.6.2/components/stdproc/orbit/orbit2sch/src/. 

cp isce 2.2.0/components/stdproc/orbit/orbit2sch/src/orbit2sch.F isce2-2.6.2/components/stdproc/orbit/orbit2sch/src/.

mv isce2-2.6.2/components/stdproc/orbit/sch2orbit/src/SConscript isce2-2.6.2/components/stdproc/orbit/sch2orbit/src/SConscript_v2.6.2

cp  isce-2.2.0/components/stdproc/orbit/sch2orbit/src/SConscript isce2-2.6.2/components/stdproc/orbit/sch2orbit/src/.

cp isce-2.2.0/components/stdproc/orbit/sch2orbit/src/sch2orbit.F isce2-2.6.2/components/stdproc/orbit/sch2orbit/src/.**

Is there any reason why these sch2orbit.F and orbit2sch.F modules were not completely migrated through the newer versions of ISCE?

Thanks
Francisco

**Comment by piyushrpt** [comment]
There is not enough RSAT2 data out there in public - so its hard to test against it for changes. Number of issues with it:

1. RSAT2 data gets shipped often with insufficient orbit vectors to span the image and beyond for reliable interpolation. This is similar to what S1 was in its early days. What is in the XML is poor quality state vectors.
2. So, we used to extend the orbit with whatever data they gave us and those old modules were used for it. There is no other workflow that needs it. Even with this, quality was often questionable because the state vectors were interpolated poor quality ones to begin with.
3. Then we found out that RSAT2 comes with precise orbit files which we support. If you have access to these - ISCE will happily read those and these work well. 

Unless you have access to these precise orbits  and you still want to force it through ISCE processing, you might have to hack the sensor and turn off the extension at the risk of failing elsewhere in orbit interpolation due to insufficient orbits in the xml files.

**Reply by domino20** [reply]
hey, 

what if the orbit quality is quite low, can stripmapApp generate the interferogram correctly? I am processing some data, and when it comes to the step runRefineSecondaryTiming(refine geometry offsets based on cross correlation), the "Bad match at level 1"  keeps coming out for almost all the lines, and get around 2pixels offset for range and 3 pixels for azimuth(guess too large, because usuallly i get results smaller than 1). The final interferogram result is also not good, some unexpected phase jump and ramp are shown.  Do you have any suggestions for this.

**Reply by fdelgadodelapuente** [reply]
Despite the low accuracy of the state vectors, I've never have any issue processing RADARSAT-2 data, except for the ramps. Can you share a screenshot of your interferogram?

**Comment by fdelgadodelapuente** [comment]
Hi Piyush,

Thanks a lot for the detailed explanation. I've processed a bunch of RSAT2 time series from the Wide Ultra Fine beams (2 m/pixel) like the next one, and I've never seen large orbital ramps that might be indicative of poor quality state vectors. 

![ifg_20220105_20221231](https://user-images.githubusercontent.com/23100180/216213456-ff1302c3-8e4f-41e7-8d10-178a99600804.png)

Since the interferograms look fine, can I just comment the orbit extender module in  isce2-2.6.2/components/isceobj/Sensor/Radarsat2.py as asked in the following code?

    def extractOrbit(self):
        '''
        Extract the orbit state vectors from the XML file.
        '''

        # Initialize orbit objects
        # Read into temp orbit first.
        # Radarsat 2 needs orbit extensions.
        tempOrbit = Orbit()

        self.frame.getOrbit().setOrbitSource('Header: ' + self.frame.getOrbit().getOrbitSource())
        stateVectors = self.product.sourceAttributes.orbitAndAttitude.orbitInformation.stateVectors
        for i in range(len(stateVectors)):
            position = [stateVectors[i].xPosition, stateVectors[i].yPosition, stateVectors[i].zPosition]
            velocity = [stateVectors[i].xVelocity, stateVectors[i].yVelocity, stateVectors[i].zVelocity]
            vec = StateVector()
            vec.setTime(stateVectors[i].timeStamp)
            vec.setPosition(position)
            vec.setVelocity(velocity)
            tempOrbit.addStateVector(vec)

        planet = self.frame.instrument.platform.planet
        orbExt = OrbitExtender(planet=planet)                       #####maybe comment this if accurate state vector are available???
        orbExt.configure()                                                         #####maybe comment this if accurate state vector are available???
        newOrb = orbExt.extendOrbit(tempOrbit)                  #####maybe comment this if accurate state vector are available???

        for sv in newOrb:                                                            #####maybe comment this if accurate state vector are available???
            self.frame.getOrbit().addStateVector(sv)                 #####maybe comment this if accurate state vector are available???

        print('Successfully read state vectors from product XML')


Francisco

**Reply by piyushrpt** [reply]
Yeah the orbit extender is the problem. If you are confident, go for it. Things may have changed with RSAT2 as well since 2019.

**Comment by fdelgadodelapuente** [comment]
Thanks Piyush for the suggestion. I commented

        #####planet = self.frame.instrument.platform.planet
        #####orbExt = OrbitExtender(planet=planet)
        #####orbExt.configure()
        #####newOrb = orbExt.extendOrbit(tempOrbit)

        #####for sv in newOrb:
        #####    self.frame.getOrbit().addStateVector(sv)

and replaced them with 

        for sv in tempOrbit:
            self.frame.getOrbit().addStateVector(sv)

It can process the data without having to add those modules from 2.2.0 and without the need to use the orbit extender.

Francisco

**Reply by ssharifi1994** [reply]
Hello Francisco,

I know it's a year apart but do you happen to remember if you ran into any issue when running stripmapApp on RS2 scenes? I made the changes you suggested here and the program runs through but the interferogram is empty! The console doesn't say anything is wrong when processing. Did you incorporate any special properties in the XML file? Mine is the basic one with referring to the reference and secondary scenes.

**Reply by fdelgadodelapuente** [reply]
Hi ssharifi1994

All my experience with RADARSAT-2 data is from the Wide Fine and Wide Ultra Fine beams. I've processed several interferorgams with stripmapApp and I've never had issues except those related to the orbit extender. The only specific thing is that you need to use the Doppler centroid provided by MDA.

	<property name="reference doppler method">useDEFAULT</property>
	<property name="secondary doppler method">useDEFAULT</property>
